Unable to build from source

**Relevant system information:**
 - OS: Fedora 33
 - PostgreSQL version 12.5
 - TimescaleDB version 2.0.0
 - Installation method: "source"

**Describe the bug**
I get the error below when trying to build it from source.

1. clone and checkout latest release;
2. run:
    ```
    ./bootstrap -DREGRESS_CHECKS=OFF
    cd build && gmake
    ```

```
[ 33%] Building C object src/CMakeFiles/timescaledb.dir/jsonb_utils.c.o
[ 34%] Building C object src/CMakeFiles/timescaledb.dir/license_guc.c.o
[ 34%] Building C object src/CMakeFiles/timescaledb.dir/partitioning.c.o
In file included from /usr/pgsql-12/include/server/postgres.h:46,
                 from /root/rpmbuild/BUILD/timescaledb/src/partitioning.c:6:
/root/rpmbuild/BUILD/timescaledb/src/partitioning.c: In function 'ts_partitioning_info_create':
/root/rpmbuild/BUILD/timescaledb/src/partitioning.c:187:2: error: 'strncpy' specified bound 64 equals destination size [-Werror=stringop-truncation]
  187 |  StrNCpy(pinfo->partfunc.name, partfunc, NAMEDATALEN);
      |  ^~~~~~~
/root/rpmbuild/BUILD/timescaledb/src/partitioning.c:188:2: error: 'strncpy' specified bound 64 equals destination size [-Werror=stringop-truncation]
  188 |  StrNCpy(pinfo->column, partcol, NAMEDATALEN);
      |  ^~~~~~~
/root/rpmbuild/BUILD/timescaledb/src/partitioning.c:196:2: error: 'strncpy' specified bound 64 equals destination size [-Werror=stringop-truncation]
  196 |  StrNCpy(pinfo->partfunc.schema, schema, NAMEDATALEN);
      |  ^~~~~~~
cc1: all warnings being treated as errors
gmake[2]: *** [src/CMakeFiles/timescaledb.dir/build.make:628: src/CMakeFiles/timescaledb.dir/partitioning.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1011: src/CMakeFiles/timescaledb.dir/all] Error 2
gmake: *** [Makefile:149: all] Error 2
error: Bad exit status from /var/tmp/rpm-tmp.u8LEud (%build)


RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.u8LEud (%build)
```

For me it fails with:
```
[ 25%] Building C object src/CMakeFiles/timescaledb.dir/dimension_vector.c.o
/home/akukushkin/git/timescaledb/src/dimension_vector.c: In function ‘ts_dimension_vec_add_slice’:
/home/akukushkin/git/timescaledb/src/dimension_vector.c:41:18: error: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Werror=strict-overflow]
  if (vec != NULL && vec->capacity >= new_capacity)
      ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/akukushkin/git/timescaledb/src/dimension_vector.c: In function ‘ts_dimension_vec_add_unique_slice’:
/home/akukushkin/git/timescaledb/src/dimension_vector.c:41:18: error: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Werror=strict-overflow]
  if (vec != NULL && vec->capacity >= new_capacity)
      ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/akukushkin/git/timescaledb/src/dimension_vector.c: In function ‘ts_dimension_vec_add_slice_sort’:
/home/akukushkin/git/timescaledb/src/dimension_vector.c:41:18: error: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Werror=strict-overflow]
  if (vec != NULL && vec->capacity >= new_capacity)
      ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
src/CMakeFiles/timescaledb.dir/build.make:614: recipe for target 'src/CMakeFiles/timescaledb.dir/dimension_vector.c.o' failed
```

It is sort of strange to have defaults that most likely would fail.

Another problem is that due to the https://github.com/timescale/timescaledb/commit/c90397fd6a0cfb5cc6487207424eeac5e6920c52 the bootstrap script fails on ubuntu 18.04 because there is only cmake 3.10 available, but perfectly works after manually changing requirement in the CMakeLists.txt.

--- a/.github/workflows/coverity.yaml
+++ b/.github/workflows/coverity.yaml
@@ -55,7 +55,7 @@ jobs:
     - name: Build TimescaleDB
       run: |
         PATH="$GITHUB_WORKSPACE/coverity/bin:$PATH"
-        ./bootstrap -DCMAKE_BUILD_TYPE=Release -DPG_SOURCE_DIR=~/$PG_SRC_DIR -DPG_PATH=~/$PG_INSTALL_DIR -DWARNINGS_AS_ERRORS=OFF
+        ./bootstrap -DCMAKE_BUILD_TYPE=Release -DPG_SOURCE_DIR=~/$PG_SRC_DIR -DPG_PATH=~/$PG_INSTALL_DIR
         cov-build --dir cov-int make -C build
 
     - name: Upload report

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -90,8 +90,15 @@ if (NOT CMAKE_C_COMPILER_ID IN_LIST SUPPORTED_COMPILERS)
    message(FATAL_ERROR "Unsupported compiler ${CMAKE_C_COMPILER_ID}. Supported compilers are: ${SUPPORTED_COMPILERS}")
 endif ()
 
-# Option to treat warnings as errors when compiling (default on)
-option(WARNINGS_AS_ERRORS "Make compiler warnings into errors (default ON)" ON)
+# Option to treat warnings as errors when compiling (default on for
+# debug builds, off for all other build types)
+if (CMAKE_BUILD_TYPE STREQUAL Debug)
+  message(STATUS "CMAKE_BUILD_TYPE matches Debug")
+  option(WARNINGS_AS_ERRORS "Make compiler warnings into errors (default ON)" ON)
+else()
+  message(STATUS "CMAKE_BUILD_TYPE does not match Debug")
+  option(WARNINGS_AS_ERRORS "Make compiler warnings into errors (default ON)" OFF)
+endif()
 
 if (WARNINGS_AS_ERRORS)
   if (CMAKE_C_COMPILER_ID MATCHES "GNU|Clang|AppleClang")
@@ -116,7 +123,8 @@ if(CMAKE_C_COMPILER_ID MATCHES "GNU|AppleClang|Clang")
   # These flags are supported on all compilers.
   add_compile_options(
     -Wempty-body -Wvla -Wall -Wmissing-prototypes -Wpointer-arith
-    -Werror=vla -Wendif-labels -fno-strict-aliasing -fno-omit-frame-pointer)
+    -Werror=vla -Wendif-labels -Wno-stringop-truncation
+    -fno-strict-aliasing -fno-omit-frame-pointer)
 
   # These flags are just supported on some of the compilers, so we
   # check them before adding them.

--- a/scripts/gh_matrix_builder.py
+++ b/scripts/gh_matrix_builder.py
@@ -35,12 +35,18 @@ m = {"include": [],}
 def build_debug_config(overrides):
   # llvm version and clang versions must match otherwise
   # there will be build errors this is true even when compiling
-  # with gcc as clang is used to compile the llvm parts
+  # with gcc as clang is used to compile the llvm parts.
+  #
+  # Strictly speaking, WARNINGS_AS_ERRORS=ON is not needed here, but
+  # we add it as a precation. Intention is to have at least one
+  # release and one debug build with WARNINGS_AS_ERRORS=ON so that we
+  # capture warnings generated due to changes in the code base or the
+  # compiler.
   base_config = dict({
     "name": "Debug",
     "build_type": "Debug",
     "pg_build_args": "--enable-debug --enable-cassert",
-    "tsdb_build_args": "-DCODECOVERAGE=ON",
+    "tsdb_build_args": "-DCODECOVERAGE=ON -DWARNINGS_AS_ERRORS=ON",
     "installcheck_args": "IGNORES='bgw_db_scheduler'",
     "coverage": True,
     "llvm_config": "llvm-config-9",
@@ -52,13 +58,17 @@ def build_debug_config(overrides):
   base_config.update(overrides)
   return base_config
 
+# We build this release configuration with WARNINGS_AS_ERRORS=ON to
+# make sure that we can build with -Werrors even for release
+# builds. This will capture some cases where warnings are generated
+# for release builds but not for debug builds.
 def build_release_config(overrides):
   base_config = build_debug_config({})
   release_config = dict({
     "name": "Release",
     "build_type": "Release",
     "pg_build_args": "",
-    "tsdb_build_args": "-DWARNINGS_AS_ERRORS=OFF",
+    "tsdb_build_args": "-DWARNINGS_AS_ERRORS=ON",
     "coverage": False,
   })
   base_config.update(release_config)
@@ -70,7 +80,7 @@ def build_apache_config(overrides):
   apache_config = dict({
     "name": "ApacheOnly",
     "build_type": "Release",
-    "tsdb_build_args": "-DAPACHE_ONLY=1 -DWARNINGS_AS_ERRORS=OFF",
+    "tsdb_build_args": "-DAPACHE_ONLY=1",
     "pg_build_args": "",
     "coverage": False,
   })

